README
NAME
    MooseX::ExpiredAttribute - Expired and auto rebuilded attributes in
    Moose objects

SYNOPSIS
        package MyClass;

        use Moose;
        use MooseX::ExpiredAttribute;

        has 'config' => (
            traits  => [ qw( Expired ) ],
            is      => 'rw',
            isa     => 'HashRef',
            expires => 5.5,
            lazy    => 1,
            builder => '_build_config',
        );

        sub _build_config {
            ... # open config file, read and hash forming
        }

        package main;

        my $prog = MyClass->new;

        $prog->config;      # First calling - here read file and hash of configs
        sleep 2;            # only 2 seconds elapsed
        $prog->config;      # Here no read - only attribute value returning
        sleep 4;            # elapsed ~ 6 seconds from first calling of builder - more than 5.5 'expires' option
        $prog->config;      # Here new read file and new hash of configs bacause old value has been expired
        sleep 3;            # only 3 seconds elapsed from rebuilding
        $prog->config;      # Here no read - only attribute value returning
        ...

    and even by this way:

        package MyRole;

        use Moose::Role;
        use MooseX::ExpiredAttribute;

        has 'config' => (
            traits  => [ qw( Expired ) ],
            is      => 'rw',
            isa     => 'HashRef',
            expires => 5.5,
            lazy    => 1,
            builder => '_build_config',
        );

        sub _build_config {
            ... # open config file, read and hash forming
        }

        package MyClass;

        with 'MyRole';

        has 'foo' => ( is => 'rw' );

        package main;

        my $prog = MyClass->new;

        $prog->config;      # First calling - here read file and hash of configs
        ...

DESCRIPTION
    This module allow to create expired attributes with auto-rebuilding
    feature after elapsed time. Goal of module is attrubutes which can be
    able to have the time-varying value. For example some configs can be
    changed by user during program runtime and wished to be reread by
    program every one minute for example. All that is required from you to
    add trait to attribute and to add 'expires' option (in fractal seconds).
    An attribute should have a builder too!

SEE ALSO
    MooseX::ExpiredAttribute::Role::Object
    MooseX::ExpiredAttribute::Role::Meta::Attribute

AUTHOR
    This module has been written by Perlover <perlover@perlover.com>

LICENSE
    This module is free software and is published under the same terms as
    Perl itself.



Changes
Revision history for perl module MooseX::ExpiredAttribute

0.022 2014-02-12
    - Few doc fixes
    - Added the required module 'Moose'

0.021 2014-02-11
    - Many small fixes (doc, git repo info, Changes file...)
    - improved test - added second expired attribute in class.

0.02 2014-02-11
    - First public release

0.01 2014-02-11
    - First release but removed from public


lib/MooseX/ExpiredAttribute/Role/Meta/Attribute.pm
package MooseX::ExpiredAttribute::Role::Meta::Attribute;

use Moose::Role;
use Moose::Util ();
use Time::HiRes ();
use MooseX::ExpiredAttribute::Role::Object ();

Moose::Util::meta_attribute_alias('Expired');

has 'expires'   => (
    is          => 'rw',
    isa         => 'Num',
);

after install_accessors => sub {
    my $self = shift;

    confess "The expired attribute '" . $self->name . "' should have builder or default coderef"
      unless $self->has_builder || $self->is_default_a_coderef;

    my $class_meta = $self->associated_class;
    my $orig_name  = $self->get_read_method;
    my $orig_meth  = $self->get_read_method_ref;

    Moose::Util::apply_all_roles( $class_meta->name, 'MooseX::ExpiredAttribute::Role::Object' )
      unless ( $class_meta->does_role( 'MooseX::ExpiredAttribute::Role::Object' ) );

    $class_meta->add_before_method_modifier( $orig_name, sub { _before_expired_accessor( $self, $orig_name, @_ ) } );
};

sub _before_expired_accessor {
    my ( $attr, $name, $self ) = @_;

    if ( ! exists $self->_expiration_time_for_attrs->{ $attr->name } || @_ == 4 ) {
        # If it's first time calling or it's writer - we set expired time in future
        $self->_expiration_time_for_attrs->{ $attr->name } = Time::HiRes::time() + $attr->expires;
    }
    elsif ( Time::HiRes::time() >= $self->_expiration_time_for_attrs->{ $attr->name } ) {
        # If value has expired we clear value = after 'before' modifiers Moose will call builder/default again
        $attr->clear_value( $self );
        $self->_expiration_time_for_attrs->{ $attr->name } = Time::HiRes::time() + $attr->expires;
    }
}

no Moose::Role;

1;
__END__

=pod

=head1 NAME

MooseX::ExpiredAttribute::Role::Meta::Attribute - the attached role to meta attribute objects (a trait)

This trait has alias as C<Expired>

=head1 DESCRIPTION

This role should be attached as I<trait> to attributes which should be have I<expired> capability.

=head1 EXAMPLE

    use MooseX::ExpiredAttribute;

    has 'config' => (
        traits  => [ qw( Expired ) ],
        is      => 'rw',
        isa     => 'HashRef',
        expires => 5.5,
        lazy    => 1,
        builder => '_build_config',
    );

Please read L<MooseX::ExpiredAttribute/SYNOPSIS> for more examples and how to use.

=head1 SEE ALSO

=over

=item L<MooseX::ExpiredAttribute>

=item L<MooseX::ExpiredAttribute::Role::Object>

=back

=head1 AUTHOR

This module has been written by Perlover <perlover@perlover.com>

=head1 LICENSE

This module is free software and is published under the same terms as Perl
itself.

=cut


lib/MooseX/ExpiredAttribute.pm
package MooseX::ExpiredAttribute;

use Moose ();

use MooseX::ExpiredAttribute::Role::Meta::Attribute;
use MooseX::ExpiredAttribute::Role::Object;

$MooseX::ExpiredAttribute::VERSION = 0.022;

1;

__END__

=pod

=head1 NAME

MooseX::ExpiredAttribute - Expired and auto rebuilded attributes in Moose objects

=head1 SYNOPSIS

    package MyClass;

    use Moose;
    use MooseX::ExpiredAttribute;

    has 'config' => (
        traits  => [ qw( Expired ) ],
        is      => 'rw',
        isa     => 'HashRef',
        expires => 5.5,
        lazy    => 1,
        builder => '_build_config',
    );

    sub _build_config {
        ... # open config file, read and hash forming
    }

    package main;

    my $prog = MyClass->new;

    $prog->config;      # The first calling - here read file and make hash of configs
    sleep 2;            # only 2 seconds elapsed...
    $prog->config;      # ... there is no calling of builder again - only attribute value returning
    sleep 4;            # elapsed ~ 6 seconds from first calling of builder - more than 5.5 seconds elapsed (the 'expires' option)
    $prog->config;      # ... there is new calling of builder - rereading config file again
    sleep 3;            # only 3 seconds elapsed from rebuilding...
    $prog->config;      # ... only old value is returned
    ...

and even by this way:

    package MyRole;

    use Moose::Role;
    use MooseX::ExpiredAttribute;

    has 'config' => (
        traits   => [ qw( Expired ) ],
        is       => 'rw',
        isa      => 'HashRef',
        expires  => 5.5,
        lazy     => 1,
        builder  => '_build_config',
    );

    sub _build_config {
        ... # open config file, read and hash forming
    }

    package MyClass;

    with 'MyRole';

    has 'foo' => ( is => 'rw' );

    package main;

    my $prog = MyClass->new;

    $prog->config;      # First calling - here read file and hash of configs
    ...

=head1 DESCRIPTION

This module allows to create expired attributes with auto-rebuilding feature
after elapsed time. The goal of module is attrubutes which can be able to have
the time-varying value. For example some configs can be changed by user during
program runtime and wished to be reread by program every one minute for example.
All that is required from you to add the trait to an attribute and to add the
'expires' option (may be fractal seconds). An attribute should have a builder
too!

=head1 SEE ALSO

=over

=item L<MooseX::ExpiredAttribute::Role::Object>

=item L<MooseX::ExpiredAttribute::Role::Meta::Attribute>

=back

=head1 AUTHOR

This module has been written by Perlover <perlover@perlover.com>

=head1 LICENSE

This module is free software and is published under the same terms as Perl
itself.

=cut
